File: mobile/apps/photos/plugins/ente_photos_platform/android/src/main/kotlin/io/ente/photos/platform/flutter/MemoryExportChannelAdapter.kt
package io.ente.photos.platform.flutter

import android.content.Context
import androidx.media3.transformer.Transformer
import io.ente.photos.platform.memoryexport.MemoryExportService
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.io.File

internal class MemoryExportChannelAdapter : MethodChannel.MethodCallHandler, EventChannel.StreamHandler {
    private lateinit var channel: MethodChannel
    private lateinit var progressChannel: EventChannel
    private var progressSink: EventChannel.EventSink? = null
    private lateinit var context: Context
    private var transformer: Transformer? = null
    private var pendingResult: MethodChannel.Result? = null

    fun attach(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        channel = MethodChannel(binding.binaryMessenger, "io.ente.photos.platform/memory_export")
        progressChannel = EventChannel(binding.binaryMessenger, "io.ente.photos.platform/memory_export/progress")
        channel.setMethodCallHandler(this)
        progressChannel.setStreamHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        if (call.method == "cancelExport") {
            try {
                cancelExport()
                result.success(null)
            } catch (error: Exception) {
                result.error("export_failed", error.message, null)
            }
            return
        }
        if (call.method != "exportMemory") {
            result.notImplemented()
            return
        }
        val input = call.argument<List<String>>("input")!!
        val output = call.argument<String>("output")!!
        if (pendingResult != null) {
            result.error("export_failed", "A memory export is already in progress", null)
            return
        }
        pendingResult = result
        try {
            transformer = MemoryExportService.exportMemory(
                context,
                input,
                File(output),
                onProgress = { progress ->
                    progressSink?.success(progress.coerceIn(0.0, 1.0))
                },
                onCompleted = {
                    transformer = null
                    pendingResult = null
                    result.success(null)
                },
                onError = { error ->
                    transformer = null
                    pendingResult = null
                    result.error("export_failed", error.message, null)
                }
            )
        } catch (error: Exception) {
            pendingResult = null
            result.error("export_failed", error.message, null)
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink) {
        progressSink = events
    }

    override fun onCancel(arguments: Any?) {
        progressSink = null
    }

    fun cancelExport() {
        transformer?.cancel()
        transformer = null
        pendingResult?.error("export_cancelled", "Memory export cancelled", null)
        pendingResult = null
    }

    fun detach() {
        cancelExport()
        channel.setMethodCallHandler(null)
        progressChannel.setStreamHandler(null)
        progressSink = null
    }
}

File: mobile/apps/photos/plugins/ente_photos_platform/ios/Flutter/MemoryExportChannelAdapter.swift
@preconcurrency import Flutter
import Foundation

@MainActor
final class MemoryExportChannelAdapter: NSObject, @preconcurrency FlutterStreamHandler {
    private let channel: FlutterMethodChannel
    private let progressChannel: FlutterEventChannel
    private var progressSink: FlutterEventSink?

    init(registrar: FlutterPluginRegistrar) {
        channel = FlutterMethodChannel(
            name: "io.ente.photos.platform/memory_export",
            binaryMessenger: registrar.messenger()
        )
        progressChannel = FlutterEventChannel(
            name: "io.ente.photos.platform/memory_export/progress",
            binaryMessenger: registrar.messenger()
        )
        super.init()
        progressChannel.setStreamHandler(self)
        channel.setMethodCallHandler { [weak self] call, result in
            if call.method == "cancelExport" {
                result(nil)
                return
            }
            guard call.method == "exportMemory" else {
                result(FlutterMethodNotImplemented)
                return
            }
            let arguments = call.arguments as! [String: Any]
            let input = arguments["input"] as! [String]
            let output = arguments["output"] as! String
            do {
                try MemoryExportService.shared.exportMemory(input: input, output: URL(fileURLWithPath: output)) { progress in
                    self?.progressSink?(min(max(progress, 0.0), 1.0))
                }
                result(nil)
            } catch MemoryExportService.ExportError.unimplemented(let message) {
                result(FlutterError(code: "unimplemented", message: message, details: nil))
            } catch {
                result(FlutterError(code: "export_failed", message: error.localizedDescription, details: nil))
            }
        }
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        progressSink = events
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        progressSink = nil
        return nil
    }

    func detach() {
        channel.setMethodCallHandler(nil)
        progressChannel.setStreamHandler(nil)
        progressSink = nil
    }
}
